Add an implementation and tests for a DataWord

whysoserious opened the pull request:
Please read the documentation in code to see what a `DataWord` is. Ping me if anything is not clear enough.

ntallar commented on src/main/scala/io/iohk/ethereum/vm/DataWord.scala:
```diff
@@ -0,0 +1,105 @@
+  override def equals(that: Any): Boolean =
+    that match {
+      case that: DataWord =>
+        that.isInstanceOf[DataWord] && this.hashCode == that.hashCode
```
Minor detail: Is `isInstanceOf[DataWord]`necessary in this case? I'm not sure if doing the pattern matching does the same type check.

LukasGasior1 replied:
also shouldn't it be `this.n == that.n` afair two non equal objects can share the same hash code

ntallar commented on src/main/scala/io/iohk/ethereum/vm/DataWord.scala:
```diff
@@ -0,0 +1,105 @@
+
+  def floatValue(): Float = longValue().toFloat
+
+  def intValue(): Int = value.foldLeft(0) { (n, b) => (n << 8) + (b & 0xff) }
```
Is there any reason why you obtain the int value from the `ByteString` `value` and not from the `BigInt` `n`? I'm not sure if maybe you intend to stop using the `n` BigInt in the future.

adamsmo replied:
what about int overflow, do we care?

adamsmo commented on src/main/scala/io/iohk/ethereum/vm/DataWord.scala:
```diff
@@ -0,0 +1,105 @@
+
+  import DataWord._
+
+  // TODO: Consider changing internal representation of a DataWord to a ByteString.
```
should this TODO be addressed?

whysoserious commented:
I addressed all your comments in the code:
- broken `.equals` method: I promise not to copy/paste from SO anymore
- `.intValue` etc.: this was required by `ScalaNumericConversions` trait which was implemented by `DataWord` which didn't make sense at all :)

adamsmo commented on src/main/scala/io/iohk/ethereum/vm/DataWord.scala:
```diff
@@ -0,0 +1,90 @@
+
+/** Stores 256 bit words and adds a few convenience methods on them.
+ *  Internally a word is stored as a BigInt. */
+class DataWord private (private val n: BigInt) {
```
can this be case class and that way we can avoid adding our own `equals` and `hashCode` ?

rtkaczyk replied:
No, it shouldn't be. We want to have our own `equals` and we don't want to expose the underlying data type - it may change in the future.

rtkaczyk commented on src/main/scala/io/iohk/ethereum/vm/DataWord.scala:
```diff
@@ -0,0 +1,90 @@
+
+  val MaxWord: BigInt = BigInt(2).pow(MaxLength * 8) - 1
+
+  val Zeros: ByteString = ByteString(Array.fill[Byte](MaxLength)(0))
```
Let's make this private. Doesn't make much sense as part of public API

rtkaczyk commented on src/main/scala/io/iohk/ethereum/vm/DataWord.scala:
```diff
@@ -0,0 +1,90 @@
+    apply(BigInt(num.toLong(n)))
+  }
+
+  def fixBigInt(n: BigInt): BigInt = {
```
Also private.

rtkaczyk commented on src/main/scala/io/iohk/ethereum/vm/DataWord.scala:
```diff
@@ -0,0 +1,90 @@
+    } else if (n < 0) {
+      n % MaxWord + MaxWord
+    } else {
+      n % MaxWord
```
This is actually incorrect (yes, I know I did that 😛 ). Let's rename `MaxWord` to `Modulus`, increment it and privatise (because it's a `BigInt` not a `DataWord`).

Also we can have it written more consisely: `(n % Modulus + Modulus) % Modulus), though the `if`/`else` version is more efficient - so just noting that. Kudos for spotting the `n == -MaxWord` case.

rtkaczyk commented on src/main/scala/io/iohk/ethereum/vm/DataWord.scala:
```diff
@@ -0,0 +1,90 @@
+
+  def ^(that: DataWord): DataWord = DataWord(this.n ^ that.n)
+
+  def unary_-(): DataWord = DataWord(-this.n)
```
Let's not use empty `()` for such functions. Empty parens suggest side effects.

rtkaczyk commented on src/main/scala/io/iohk/ethereum/vm/DataWord.scala:
```diff
@@ -0,0 +1,90 @@
+  override def equals(that: Any): Boolean = {
+    that match {
+      case that: DataWord => this.n.equals(that.n)
+      case _ => false
```
`case other => other == n`. This way, having: `val a = DataWord(0)` I can write: `a == 0`, which is more efficient convenient and readable than: `a == DataWord(0)`.

rtkaczyk commented on src/main/scala/io/iohk/ethereum/vm/DataWord.scala:
```diff
@@ -0,0 +1,90 @@
+    }
+  }
+
+  override def hashCode(): Int = n.hashCode()
```
Remove empty parens

rtkaczyk commented on src/main/scala/io/iohk/ethereum/vm/DataWord.scala:
```diff
@@ -0,0 +1,90 @@
+
+  override def hashCode(): Int = n.hashCode()
+
+  override def toString(): String = {
```
We won't use `toString` for anything other than debugging, and for that I find this the most useful form:

```
DataWord(16).toString == "DataWord(0x10)"
```

The format would be: `f"DataWord(0x$n%02x)"`

rtkaczyk commented:
> .intValue etc.: this was required by ScalaNumericConversions trait which was implemented by DataWord which didn't make sense at all :)

Actually, we need `intValue` for multiple reasons. But that doesn't mean we necessarily need `ScalaNumericConversions`

rtkaczyk commented:
I have pushed the improvements I described in comments. Also gave the tests a closer look - they're very good!

LukasGasior1 commented:
lgtm

rtkaczyk commented:
I would also give an LGTM, but since I've committed to it I shouldn't. @adamsmo @ntallar ?
